[WIP] Support for external thin clients 2.0

danielsuo opened the pull request:
## What do these changes do?

Creating a new pull request (old one at #2282) because I don't want to lose conversation history, but also want to have a clean slate rebased on ```0.5.0``` with raylets.

Enable a thin client (i.e., no processes besides the user's [Python] process) to connect to and use a Ray cluster that may be behind a firewall.

A **BIG** caveat: this is currently prototype code and I believe one test is failing. May the internet forever be kind to us for posting this.

Getting started:
- Run ```pip install requests flask ```
- Install ```socat```
- Run the tests ```python test/client_test.py```

Some links
- [Overview document](https://docs.google.com/document/d/1iVCrHBQeF4Xq5HqFOVSDTWLDtnnjpi0LaQEttkAI7fI/edit)
- [Task list](https://docs.google.com/document/d/1IF1OdgS8w8WgXVSg2zUAOV7IFgq24SUu2Nf3HsGHkdc/edit)
- [Some slides](https://docs.google.com/presentation/d/1HPTpLLnNT1O5EWepUBVM556mKUWqB0C53iY2tk5FiHg/edit#slide=id.g3d118c42b2_0_262)  for the visually inclined

## Related issue number

#2282

danielsuo commented:
Still a "WIP WIP" i.e., needs testing to make sure the raylet code path works.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/6831/
Test FAILed.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -251,6 +254,10 @@ def __init__(self):
         self.task_driver_id = None
 
+        # Workers typically are not external clients
+        # TODO (dsuo): should rename this so it's clear what "client" means
```
How about we use the terminology "external_client" instead of "client" throughout this PR. What do you think?

danielsuo replied:
A bit long, but it's more important to distinguish what ```client``` refers to. As long as you're ok with the following:

- ```EXTERNAL_CLIENT_MODE```
- ```external_client```
- ```ExternalClient```

I'm good!

robertnishihara replied:
Yeah, descriptive is good.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -2293,6 +2369,13 @@ def disconnect(worker=global_worker):
     worker.cached_remote_functions_and_actors = []
     worker.serialization_context_map.clear()
 
+    subprocess.call(
```
This should be part of https://github.com/ray-project/ray/blob/46351957bb633b1860b4e379e592e2cd9455600a/python/ray/services.py#L170-L188

robertnishihara commented on python/ray/worker.py:
```diff
@@ -2492,10 +2585,16 @@ def put(value, worker=global_worker):
         if worker.mode == LOCAL_MODE:
             # In LOCAL_MODE, ray.put is the identity operation.
-            return value
+            originreturn value
```
typo

robertnishihara commented on test/client_test.py:
```diff
@@ -0,0 +1,38 @@
+@pytest.fixture
+class ClientTest():
+    def tearDown(self):
+        ray.worker.cleanup()
```
We switched to `ray.shutdown`

robertnishihara commented on test/client_test.py:
```diff
@@ -0,0 +1,38 @@
+        ray.worker.cleanup()
+
+    def testBasicClient(self):
+        ray.init(driver_mode=ray.CLIENT_MODE)
```
Calling `ray.init(driver_mode=ray.CLIENT_MODE)` should not start a Ray cluster (raylet, object store, etc), right? Are those processes being started somewhere? I would just expect it to start a lightweight external client which connects to an existing gateway.

I wouldn't expect `ray.init(driver_mode=ray.CLIENT_MODE)` to make any sense without also passing in a gateway address. Or am I misunderstanding the semantics?

danielsuo replied:
These tests just make sure none of the most basic pathways are broken. If no redis address is specified, the various head node processes are started locally and the client connects to this local head node. There are some comments on this use case that say this is useless except in testing.

If it's too confusing, we can ban ```driver_mode=ray.CLIENT_MODE``` if a ```redis``` address isn't specified.

robertnishihara commented on test/client_test.py:
```diff
@@ -0,0 +1,38 @@
+
+# NOTE: These client tests run a gateway locally, which I don't think
+# makes any sense in any real-world scenario. A true Client test would
+# require multiple machines with a network configured more realistically.
```
One way to do the test on a single machine would be to first start the Ray cluster with

```
ray start --head --with-gateway ...
```

probably inside of `subprocess.Popen`, and then connect here via

```python
ray.init(gateway_address=...)
```

Some examples of this kind of thing are in e.g., https://github.com/ray-project/ray/blob/46351957bb633b1860b4e379e592e2cd9455600a/test/multi_node_test.py#L32 though note that it's ok to defer this kind of test to a subsequent PR.

robertnishihara commented on python/ray/client.py:
```diff
@@ -0,0 +1,97 @@
+            self.gateway_address,
+            self.gateway_data_port)
+        
+        if client_socket_name is not None:
```
Is there a one-to-one mapping between `Client` objects and named pipes for forwarding  info? If so, should we just create the named pipe in the constructor here?

robertnishihara commented on python/ray/gateway.py:
```diff
@@ -0,0 +1,112 @@
+from ray.utils import hex_to_binary
+
+app = Flask(__name__)
+ctx = pyarrow.default_serialization_context()
```
It's possible that we already discussed this and I forgot the relevant issues, but I think it should be possible to get away without having a serialization context in the gateway (e.g., to never have the gateway serialize or deserialize anything).

danielsuo replied:
Yep. This was just a copy over; haven't gotten to it yet.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -1633,7 +1656,10 @@ def _init(address_info=None,
             include_webui=include_webui,
-            use_raylet=use_raylet)
+            use_raylet=use_raylet,
+            with_gateway=(driver_mode == CLIENT_MODE),
```
This argument doesn't seem to have been piped through to `services.py`.

robertnishihara commented:
Responding to https://github.com/ray-project/ray/pull/2478#discussion_r205590421.

Question: When an external client connects to the cluster. Does it need both the redis address and the gateway address? Or just the redis address? Or just the gateway address?

danielsuo commented:
Right now, we just specify ```redis_address``` and assume that the gateway is on the same machine at a different port. Eventually, you'd only want to specify the gateway address and route ```redis``` via gateway.

danielsuo commented:
In fact, we could make ```redis_address``` and ```gateway_address``` mutually exclusive options with the latter indicating we are in ```CLIENT_MODE```.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7012/
Test FAILed.

danielsuo commented:
Botched a rebase and the commit history is now polluted. Will clean up.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7016/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7015/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7018/
Test FAILed.

danielsuo commented:
@robertnishihara turns out the "minimum" implementation has resisted some of the step-by-step breakdown we discussed. I *think* once I resolve the serialization issues, we should be good for a first cut (still very rough, of course).

A side note -- the jenkins keeps failing because of a timeout. I don't think that has anything to do with the external client work, but wasn't sure if this was something common you guys ran into.

robertnishihara commented:
> In fact, we could make `redis_address` and `gateway_address` mutually exclusive options with the latter indicating we are in `CLIENT_MODE`. >

That sounds best. `CLIENT_MODE` can just be inferred.

robertnishihara commented:
@danielsuo yeah, the Jenkins timeout is unrelated. It's fairly common at the moment unfortunately.

Can you remind me what the current serialization issue that you're referring to is?

danielsuo commented:
Two separate things:

1. Serializing ```numpy``` arrays is failing, which is surprising to me; something is getting lost during in the HTTP connection, which is odd since it's supposed to just dump bytes. Shouldn't be hard to get to the bottom of.

2. We can get rid of the extra deserialization / serialization in ```Gateway```, but ```ExternalClient``` still needs to serialize ```put```s and deserialize ```get```s. The way we've done it now is ugly, but we can continue to discuss.

robertnishihara commented:
For 2, each `ray.put` and `ray.get` conceptually consists of two parts: serialization/deserialization and copying to/from the object store. We combine them for performance reasons, but it may make sense to do the "serialization/deserialization" part on the `ExternalClient` and the "copy to/from the object store" on the `GateWay`. Could something like that work?

The serialization part is done with, e.g.,

```python
serialized_obj = ray.pyarrow.serialize(obj).to_buffer().to_pybytes()
ray.pyarrow.deserialize(serialized_obj)
```

The copy to/from the object store can be done with

```
object_id = ray.pyarrow.plasma.ObjectID(np.random.bytes(20))
buff = ray.worker.global_worker.plasma_client.create(object_id, len(serialized_obj))

# Do the copy. There's almost certainly a better way of doing this of course.
for i in range(len(serialized_obj)):
    buff[i] = serialized_obj[i]

ray.worker.global_worker.plasma_client.seal(object_id)

###### Get
ray.worker.global_worker.plasma_client.get_buffers([object_id])[0]
```

danielsuo commented:
Ah, my earlier comment was confusing; what you just described is exactly how we're changing the code. I was just looking into the "better way of doing [byte array copy]".

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7101/
Test FAILed.

danielsuo commented:
Annoying; can't read a ```SpooledTemporaryFile``` (i.e., what Flask receives data as) into a memoryview via ```readinto```. Will leave a comment to use readinto if / when we move away from Flask.

danielsuo commented:
@robertnishihara I think we're getting close to initial merge. We can get /put all serializable objects (though no support for multiple ```ObjectID```s in a single call yet). The main thing to do is set up the multi-node testing.

Of course, let me know if you disagree / feel there is something we should also get in. Given that this could be a much more involved effort than what we've done so far, perhaps it's worth charting  out where we want to go from here.

danielsuo commented:
To summarize the conversations so far, I believe the only unaddressed comments are:

- Forwarding exceptions and log messages (last comment from #2282)
- The gateway is somewhat cumbersome to use. First, it's yet another process on the head node (or wherever it's run), it assumes it's running on the same machine as the redis instance, and adds two additional ports that need to be opened (one for ```socat``` to proxy commands and one to send / receive data from a REST web server)

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7106/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7111/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7110/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7142/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7143/
Test PASSed.

danielsuo commented:
@robertnishihara Keep getting seemingly unrelated errors in Travis (e.g., ```ActorsWithGPUs.testActorMultipleGPUsFromMultipleTasks``` keeps failing under Linux, but not MacOS). I noticed that many PRs and builds fail in Travis; trying to get all tests to pass, but perhaps this is not a good goal?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7158/
Test FAILed.

danielsuo commented:
@robertnishihara just wanted to bump this given all that's going on in Rayland. A few things:

1. We were able to integrate this PR into our own workflow successfully
2. Wondering if we should figure out a roadmap for this feature in general since this is just a first cut
3. Tests that appear unrelated to this PR are failing (mentioned above)

robertnishihara commented:
@danielsuo ah apologies for the delay, I'll give it a try now!

robertnishihara commented:
@danielsuo still trying things out. Will need to continue tomorrow.

@alexbao @ericl @suquark you may be interested in taking a look at this PR.

danielsuo commented:
@robertnishihara no worries. We're not in a rush since the prototype is working for our purposes now. Would be interested in rewriting this as @suquark works on the ```worker.py``` refactor (hence our interest in how that's done).

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7347/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7351/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7350/
Test FAILed.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -39,12 +39,14 @@
     random_string,
     thread_safe_client,
 )
+from ray.external_client import ExternalClient
```
If we don't do this import, then `import ray` will still succeed even if `requests` is not installed. I suppose `requests` is probably pretty lightweight, so that may not be an issue, but it seems ok to defer this import until we actually instantiate the `ExternalClient`.

danielsuo replied:
Sounds good to me!

danielsuo replied:
I can create an optional set of dependencies for external client and add to the documentation.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -1683,6 +1706,20 @@ def _init(address_info=None,
+            "redis_address": address_info["redis_address"],
+            "gateway_socat_port": gateway_socat_port,
+            "gateway_data_port": gateway_data_port,
+            "raylet_socket_name": address_info["raylet_socket_names"][0]
```
This is failing for me when I try to connect to an external client to a remote machine, because `address_info` which is returned by `get_address_info_from_redis` consists of the address info for the processes on the local machine, but there are none on the local machine.

This does work for me if I attach the external client from the same node that `ray start` was run on.

Am I using it incorrectly or something?

robertnishihara replied:
I thought the issue might be that I'm trying this on AWS, where nodes have distinct public and private IP addresses, but I don't think that's the issue.

danielsuo replied:
Yeah, something fishy is going on here. Will have a look. Tests are passing for me, so need to reproduce this.

danielsuo commented:
Hm the function hang and issue with EC2 are weird. I’ll have a look tomorrow once I’ve shipped off the kids. Almost certainly missed something.

danielsuo commented on doc/source/using-ray-from-outside-the-cluster.rst:
```diff
@@ -0,0 +1,32 @@
+
+.. code-block:: bash
+  
+  ray start --head --with-gateway --use-raylet --redis-port=21216
```
Sounds good. I used a different high-numbered port since were accessing remotely and didn’t want to open some better-known port.

robertnishihara commented:
Ok, I'll try again later tonight. I need to double check that I was compiled everything correctly and on the right branch..

danielsuo commented:
Actually, hold your beer. I just checked out a fresh copy and created a new
environment and I’m failing. I’ve done something wrong; let me figure out
first before you spend more time.

On Wed, Aug 8, 2018 at 8:38 PM Robert Nishihara <notifications@github.com>
wrote:

> Ok, I'll try again later tonight. I need to double check that I was
> compiled everything correctly and on the right branch..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ray-project/ray/pull/2478#issuecomment-411599338>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACkV1obEOwil2YWkZJn7Jf6-MI6FP82Gks5uO4R-gaJpZM4VgZjt>
> .
>

robertnishihara commented:
Ok, sounds good.

danielsuo commented:
Turns out my original testing was done in the old, non-```raylet``` code path, which is why it was working. Sad. Getting everything to play nicely with ```raylet``` has been a little trickier; will spill over to next week, unfortunately.

robertnishihara commented:
Got it, good to know.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/7719/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/8213/
Test FAILed.

ericl commented:
[closing stale issue]
